devito/dse/rewriters.py

```diff
@@ -406,6 +408,40 @@ def _extract_sum_of_products(self, cluster, template, **kwargs):
 
         return cluster.rebuild(processed)
 
+    @dse_pass
+    def _skewing(self, cluster, template, **kwargs):
+        """
+        Skew the accesses along the time Dimension.
+        """
+
+        MAX_SKEW_FACTOR = 8
+        """
+        The maximum skew factor applied for the skewing pass.
+        """
+
+        skew_factor = 2  # TOFIX
+
+        if skew_factor not in range(0, MAX_SKEW_FACTOR):
+            raise ValueError("skew_factor must be less than %d" % MAX_SKEW_FACTOR)
+
+        skew_dim, mapper, intervals = None, {}, []
+
+        for i in cluster.ispace.intervals:
+            if i.dim.is_Time:
+                intervals.append(Interval(i.dim, 0, 0))
+                skew_dim = i.dim
+            else:
+                mapper[i.dim] = i.dim + skew_factor*skew_dim
+                intervals.append(Interval(i.dim, -skew_factor*skew_dim,
+                                          -skew_factor*skew_dim))
+
+        # TODO Can I avoid 2 rebuilds?
+        processed = xreplace_indices(cluster.exprs, mapper)
+        cluster.rebuild(processed)
```

Comment:
What is the purpose of this rebuild? 
The effect of this line is 0. A rebuild returns something, which you're dropping on the floor, so... it's like lines 439-440 are not there

Reply:
Given the TODO comment and what you're doing here, my feeling is that immutability vs mutability isn't completely clear yet. 

With immutability, we *rebuild* objects, instead of changing their state. Rebuilding means creating a complete new object (in which potentially only some attributes have changed)

Reply:
> What is the purpose of this rebuild?
> The effect of this line is 0. A rebuild returns something, which you're dropping on the floor, so... it's like lines 439-440 are not there

This makes sense. My mistake.